fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -157,12 +158,13 @@ where
     pub fn check_predicates(
         checked: &Checked<Tx>,
         params: &CheckPredicateParams,
+        pool: VmPool,
```

Comment:
For the synchronous predicate execution, we don't need `VmPool`; we only need one instance of `Memory`.

Reply:
Since you need to return the memory back, you could put it into the `PredicatesChecked`, but it is very unintuitive.

The best would be to pass `&mut Memory`. In this case, the interface would be clear. But it means we need to teach `Interpreter` to work with `Memory` and `&mut Memory` types a.k.a. add a new generic and trait for the `Memory`.

What do you think?

Reply:
I generally dislike adding new traits when not absolutely necessary, and here it would also incur runtime overhead, or be an inconvenitent breaking change if we add new generic parameter to the `Interpreter` type. I'll do some experiments and see what can be done. Other approach would be to make some kind of wrapper around the `Memory` that allows recycling it on drop.

That said, I don't mind too much just passing in the pool here. It likely doesn't have any performance impact, and the interface isn't much worse anyway.

Reply:
> That said, I don't mind too much just passing in the pool here. It likely doesn't have any performance impact, and the interface isn't much worse anyway.

The `VmPool` is a stronger requirement than a single `Memory` instance. For example, using the pool means that on the `fuel-core-executor` side, we now need to use a pool, while we could use one `Memory` instance. Based on the current implementation, usage of the pool means we need to use a `Mutex` in places where we don't need it. Because the executor is fully synchronous right now.

> It likely doesn't have any performance impact

Usage of the pool pulls additional dependencies that we don't need. Like mutex, adding small overhead.

> here it would also incur runtime overhead

What runtime overhead are you talking about?

> Other approach would be to make some kind of wrapper around the Memory that allows recycling it on drop.

With a new generic, it will be easy to do since the wrapper can implement our trait=)


BTW, using `&mut Memory` would avoid errors like the one below, where we forgot to recycle the memory back:

<img width="1094" alt="image" src="https://github.com/FuelLabs/fuel-vm/assets/18346821/3c2dc9b5-9799-4a24-97e4-be342a1e644f">

Reply:
Mostly done in e82441b1